File: replication/run_all.py
"""
run_all.py
==========
One-shot driver: rebuild the master panel and regenerate every table and figure
in the Charron (2013) replication, in dependency order.

    python3 run_all.py

Outputs land in ./output/ (text tables, CSVs, and PNG figures).
"""

import runpy
import os
import shutil
import subprocess

HERE = os.path.dirname(os.path.abspath(__file__))
os.chdir(HERE)

STEPS = [
    ("build_dataset.py",          "Master dyad-year panel"),
    ("table1.py",                 "Table 1  - top bias dyads"),
    ("cluster_table2_fig1.py",    "Table 2 + Fig 1  - friend groups / dendrogram"),
    ("fig2_network.py",           "Fig 2  - friend network (method 2)"),
    ("appendix3.py",              "Appendix 3  - within-group bias"),
    # Python regressions also EXPORT output/regression_data.csv (the merged,
    # N-matched frame) for R to estimate on.
    ("regressions_table3_4_fig3.py", "Tables 3-4 + Fig 3  - regressions (+ export regression_data.csv)"),
]

for script, desc in STEPS:
    print("\n" + "=" * 78)
    print(f"RUNNING {script}  --  {desc}")
    print("=" * 78)
    runpy.run_path(os.path.join(HERE, script), run_name="__main__")

# AUTHORITATIVE estimation is done in R (AER::tobit is a stable, standard Tobit;
# the Python MLE is only a cross-check).  regressions.R reads the Python-exported
# regression_data.csv and writes table3_R.txt / table4_R.txt.
print("\n" + "=" * 78)
print("RUNNING regressions.R  --  Tables 3-4 (authoritative, AER::tobit)")
print("=" * 78)
if shutil.which("Rscript"):
    subprocess.run(["Rscript", os.path.join(HERE, "regressions.R")], check=True)
else:
    print("Rscript not found on PATH -- skipping the R tables (install R + "
          "AER/sandwich/lmtest to regenerate table3_R.txt / table4_R.txt).")

print("\nAll steps complete.  See ./output/ for tables (.txt/.csv) and figures (.png).")

File: replication/run_all.R
# run_all.R
# =========
# One-shot driver for the full R pipeline (mirrors run_all.py). Regenerates the
# R master panel, all tables, all figures and the regressions, in order.
#
#     Rscript run_all.R
#
# Outputs land in ./output/ with an _R suffix so they sit alongside the Python
# outputs without clobbering them.

get_here <- function() {
  a <- commandArgs(FALSE); m <- grep("^--file=", a)
  if (length(m)) return(dirname(normalizePath(sub("^--file=", "", a[m]))))
  getwd()
}
here <- get_here()

steps <- list(
  c("build_dataset.R",       "Master dyad-year panel"),
  c("table1.R",              "Table 1  - top bias dyads"),
  c("cluster_table2_fig1.R", "Table 2 + Fig 1"),
  c("fig2_network.R",        "Fig 2 network (method 2)"),
  c("appendix3.R",           "Appendix 3"),
  c("regressions.R",         "Tables 3-4 + Fig 3")
)
for (s in steps) {
  cat("\n", strrep("=", 70), "\nRUNNING ", s[1], "  --  ", s[2], "\n",
      strrep("=", 70), "\n", sep = "")
  system2("Rscript", file.path(here, s[1]))
}
cat("\nAll R steps complete. See ./output/ for *_R tables and figures.\n")
